sympy/functions/combinatorial/numbers.py

```diff
@@ -506,49 +540,65 @@ def _calc_bernoulli(n):
-                        p, q = bernfrac(n)
-                        return Rational(int(p), int(q))
-                    case = n % 6
-                    highest_cached = cls._highest[case]
-                    if n <= highest_cached:
-                        return cls._cache[n]
-                    # To avoid excessive recursion when, say, bernoulli(1000) is
-                    # requested, calculate and cache the entire sequence ... B_988,
-                    # B_994, B_1000 in increasing order
-                    for i in range(highest_cached + 6, n + 6, 6):
-                        b = cls._calc_bernoulli(i)
-                        cls._cache[i] = b
-                        cls._highest[case] = i
-                    return b
-                # Bernoulli polynomials
-                else:
-                    n = int(n)
-                    result = [binomial(n, k)*cls(k)*sym**(n - k) for k in range(n + 1)]
-                    return Add(*result)
-            else:
-                raise ValueError("Bernoulli numbers are defined only"
-                                 " for nonnegative integer indices.")
-
-        if sym is None:
-            if n.is_odd and (n - 1).is_positive:
+    def eval(cls, n, x=None):
+        if x is S.One:
+            return cls(n)
+        elif n.is_zero:
+            return S.One
+        elif n.is_integer is False or n.is_nonnegative is False:
+            return
+        # Bernoulli numbers
+        elif x is None:
+            if n is S.One:
+                return S.Half
+            elif n.is_odd and (n-1).is_positive:
                 return S.Zero
+            elif n.is_Number:
+                n = int(n)
+                # Use mpmath for enormous Bernoulli numbers
+                if n > 500:
+                    p, q = bernfrac(n)
+                    return Rational(int(p), int(q))
+                case = n % 6
+                highest_cached = cls._highest[case]
+                if n <= highest_cached:
+                    return cls._cache[n]
+                # To avoid excessive recursion when, say, bernoulli(1000) is
+                # requested, calculate and cache the entire sequence ... B_988,
+                # B_994, B_1000 in increasing order
+                for i in range(highest_cached + 6, n + 6, 6):
+                    b = cls._calc_bernoulli(i)
+                    cls._cache[i] = b
+                    cls._highest[case] = i
+                return b
+        # Bernoulli polynomials
+        elif n is S.One:
+            return x - S.Half
+        elif n.is_Number:
+            n = int(n)
+            result = [binomial(n,k) * cls(k) * x**(n-k) for k in range(0, n+1, 2)]
+            return Add(*result) - S.Half * n * x**(n-1)
+
+    def _eval_rewrite_as_zeta(self, n, x=1, **kwargs):
+        from sympy.functions.special.zeta_functions import zeta
+        from sympy.functions.elementary.piecewise import Piecewise
+        from sympy.core.relational import Eq
```

Comment:
Do all of these imports need to be in the function like this rather than at the top of the file?

Reply:
Imports in function methods are common throughout the functions module – I see no need to break this pattern. Besides, pushing the imports up to module scope would cause circular import errors, such as with my `zeta()` and `bernoulli()` in different modules referencing each other.

Reply:
> Imports in function methods are common throughout the functions module

Ideally they wouldn't be. There is an actual runtime cost that comes from having so many delayed imports.

The pattern comes from trying to avoid circular imports but ends up being used in situations where there wouldn't even be any circular imports. I'm pretty sure that here `Eq` can just be imported at the top for example.

Reply:
Still, done. When is the earliest possible SymPy release that can include this PR?

Reply:
It will be 1.12. Probably in a few months.